global_install/hooks/pre-commit

```diff
@@ -0,0 +1,217 @@
+            '`pre-commit uninstall`'.format(config),
+        )
+
+
+def _exe():
+    with open(os.devnull, 'wb') as devnull:
+        for exe in (INSTALL_PYTHON, sys.executable):
+            try:
+                if not subprocess.call(
+                        (exe, '-c', 'import pre_commit.main'),
+                        stdout=devnull, stderr=devnull,
+                ):
+                    return (exe, '-m', 'pre_commit.main', 'run')
+            except OSError:
+                pass
+
+    if distutils.spawn.find_executable('pre-commit'):
+        return ('pre-commit', 'run')
+
+    raise FatalError(
+        '`pre-commit` not found.  Did you forget to activate your virtualenv?',
+    )
+
+
+def _rev_exists(rev):
+    return not subprocess.call(('git', 'rev-list', '--quiet', rev))
+
+
+def _pre_push(stdin):
+    remote = sys.argv[1]
+
+    opts = ()
+    for line in stdin.decode('UTF-8').splitlines():
+        _, local_sha, _, remote_sha = line.split()
+        if local_sha == Z40:
+            continue
+        elif remote_sha != Z40 and _rev_exists(remote_sha):
+            opts = ('--origin', local_sha, '--source', remote_sha)
+        else:
+            # ancestors not found in remote
+            ancestors = subprocess.check_output((
+                'git', 'rev-list', local_sha, '--topo-order', '--reverse',
+                '--not', '--remotes={}'.format(remote),
+            )).decode().strip()
+            if not ancestors:
+                continue
+            else:
+                first_ancestor = ancestors.splitlines()[0]
+                cmd = ('git', 'rev-list', '--max-parents=0', local_sha)
+                roots = set(subprocess.check_output(cmd).decode().splitlines())
+                if first_ancestor in roots:
+                    # pushing the whole tree including root commit
+                    opts = ('--all-files',)
+                else:
+                    cmd = ('git', 'rev-parse', '{}^'.format(first_ancestor))
+                    source = subprocess.check_output(cmd).decode().strip()
+                    opts = ('--origin', local_sha, '--source', source)
+
+    if opts:
+        return opts
+    else:
+        # An attempt to push an empty changeset
+        raise EarlyExit()
+
+
+def _opts(stdin, config):
+    fns = {
+        'prepare-commit-msg': lambda _: ('--commit-msg-filename', sys.argv[1]),
+        'commit-msg': lambda _: ('--commit-msg-filename', sys.argv[1]),
+        'pre-commit': lambda _: (),
+        'pre-push': _pre_push,
+    }
+    stage = HOOK_TYPE.replace('pre-', '')
+    return ('--config', config, '--hook-stage', stage) + fns[HOOK_TYPE](stdin)
+
+
+if sys.version_info < (3, 7):  # https://bugs.python.org/issue25942
+    def _subprocess_call(cmd):  # this is the python 2.7 implementation
+        return subprocess.Popen(cmd).wait()
+else:
+    _subprocess_call = subprocess.call
+
+
+def main():
+    config_path = os.path.join(top_level, ".git/config")
```

Comment:
So this is run per repo looking for alphagov in the .git/config. What runs this?

Reply:
it's a global precommit hook!